File: backend/internal/handler/v1/task.go
package v1

import (
	"context"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"

	"github.com/masahikokawai/training-go/bff-gin/backend/internal/authjwt"
	"github.com/masahikokawai/training-go/bff-gin/backend/internal/model"
	"github.com/masahikokawai/training-go/bff-gin/backend/internal/service"
)

// userLookup はTaskHandlerがJWTのclaimsからアプリ内部のユーザーを引くために必要な最小のインターフェース
// *repository.Userがこれらのメソッドを実装しているため
// 本番の配線(cmd/server/main.go)は変更不要だが、テストではfakeに差し替えられる
//
// 【セクション16.5で追加】Getはローカル(HMAC/RSA)発行のJWT用
// GetByKeycloakSub は Keycloak 発行の JWT用(resolveUserID参照)
type userLookup interface {
	GetByKeycloakSub(ctx context.Context, keycloakSub string) (*model.User, error)
	Get(ctx context.Context, id uint64) (*model.User, error)
}

// TaskHandler はREST v1(旧実装)のTaskエンドポイント
// ListだけがN+1を再現するservice.ListLegacyを呼ぶ(CONTRACT.md参照)
type TaskHandler struct {
	tasks *service.TaskService
	users userLookup
}

func NewTaskHandler(tasks *service.TaskService, users userLookup) *TaskHandler {
	return &TaskHandler{tasks: tasks, users: users}
}

// resolveUserID はJWT検証済みclaimsのsubから内部ユーザーIDを引く(gRPC v2と同じ考え方)
//
// 【セクション16.5で変更】claims.Issuerで発行元を判定し、ローカル(HMAC/RSA)発行の
// JWTはsubに内部user_idそのものが入っている(bffが発行時にそう設定する
// CONTRACT.mdセクション16.4)ためGetで直接引き、Keycloak発行のJWTは従来通り
// subがkeycloak_subなのでGetByKeycloakSubで引く
func (h *TaskHandler) resolveUserID(c *gin.Context) (uint64, bool) {
	claims, ok := authjwt.ClaimsFromContext(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return 0, false
	}

	if authjwt.IsLocalIssuer(claims.Issuer) {
		id, err := strconv.ParseUint(claims.Subject, 10, 64)
		if err != nil {
			c.JSON(http.StatusForbidden, gin.H{"error": "user_not_provisioned"})
			return 0, false
		}
		user, err := h.users.Get(c.Request.Context(), id)
		if err != nil {
			c.JSON(http.StatusForbidden, gin.H{"error": "user_not_provisioned"})
			return 0, false
		}
		return user.ID, true
	}

	user, err := h.users.GetByKeycloakSub(c.Request.Context(), claims.Subject)
	if err != nil {
		c.JSON(http.StatusForbidden, gin.H{"error": "user_not_provisioned"})
		return 0, false
	}
	return user.ID, true
}

// List は GET /internal/v1/tasks
//
// 【N+1が発生する箇所】ここが呼ぶ service.ListLegacy は、一覧取得後にタスクごとに
// ラベルを個別取得する旧実装であり、意図的にN+1を再現している(CONTRACT.md参照)
func (h *TaskHandler) List(c *gin.Context) {
	userID, ok := h.resolveUserID(c)
	if !ok {
		return
	}

	filter := service.TaskListFilterV1{
		Name:           c.Query("name"),
		SortFinishedOn: c.Query("sort"),
		Limit:          queryIntDefault(c, "limit", 20),
		Offset:         queryIntDefault(c, "offset", 0),
		LabelIDs:       parseUintListQuery(c, "label_ids"),
	}
	if s := c.Query("status"); s != "" {
		st, err := model.TaskStatusFromString(s)
		if err != nil {
			c.JSON(http.StatusUnprocessableEntity, gin.H{"error": "invalid_status"})
			return
		}
		filter.Status = &st
	}

	dtos, total, err := h.tasks.ListLegacy(c.Request.Context(), userID, filter)
	if err != nil {
		renderServiceError(c, err)
		return
	}

	tasksJSON := make([]gin.H, 0, len(dtos))
	for _, dto := range dtos {
		tasksJSON = append(tasksJSON, taskDTOToJSON(dto))
	}
	c.JSON(http.StatusOK, gin.H{
		"tasks":  tasksJSON,
		"total":  total,
		"limit":  filter.Limit,
		"offset": filter.Offset,
	})
}

func (h *TaskHandler) Get(c *gin.Context) {
	userID, ok := h.resolveUserID(c)
	if !ok {
		return
	}
	id, err := parseUintParam(c, "id")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid_id"})
		return
	}
	dto, err := h.tasks.Get(c.Request.Context(), id, userID)
	if err != nil {
		renderServiceError(c, err)
		return
	}
	c.JSON(http.StatusOK, taskDTOToJSON(dto))
}

// json タグはCONTRACT.md セクション5.1(bffのtaskV1RequestBodyと合わせたスネークケース)に従う
// bffのresty呼び出しはこのケースで送信するため、ここをcamelCaseにするとbinding:"required"が
// 常に失敗し、Create/Updateがすべて400になる(実際に統合時に発覚した不整合)
type taskRequestBody struct {
	Name        string   `json:"name" binding:"required"`
	Description *string  `json:"description"`
	Status      string   `json:"status" binding:"required"`
	FinishedOn  string   `json:"finished_on" binding:"required"`
	LabelIDs    []uint64 `json:"label_ids"`
}

func (h *TaskHandler) Create(c *gin.Context) {
	userID, ok := h.resolveUserID(c)
	if !ok {
		return
	}
	var body taskRequestBody
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid_request"})
		return
	}
	finishedOn, err := time.Parse("2006-01-02", body.FinishedOn)
	if err != nil {
		c.JSON(http.StatusUnprocessableEntity, gin.H{"error": "invalid_finished_on"})
		return
	}
	dto, err := h.tasks.Create(c.Request.Context(), userID, service.TaskInput{
		Name:        body.Name,
		Description: body.Description,
		Status:      body.Status,
		FinishedOn:  finishedOn,
		LabelIDs:    body.LabelIDs,
	})
	if err != nil {
		renderServiceError(c, err)
		return
	}
	c.JSON(http.StatusCreated, taskDTOToJSON(dto))
}

func (h *TaskHandler) Update(c *gin.Context) {
	userID, ok := h.resolveUserID(c)
	if !ok {
		return
	}
	id, err := parseUintParam(c, "id")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid_id"})
		return
	}
	var body taskRequestBody
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid_request"})
		return
	}
	finishedOn, err := time.Parse("2006-01-02", body.FinishedOn)
	if err != nil {
		c.JSON(http.StatusUnprocessableEntity, gin.H{"error": "invalid_finished_on"})
		return
	}
	dto, err := h.tasks.Update(c.Request.Context(), id, userID, service.TaskInput{
		Name:        body.Name,
		Description: body.Description,
		Status:      body.Status,
		FinishedOn:  finishedOn,
		LabelIDs:    body.LabelIDs,
	})
	if err != nil {
		renderServiceError(c, err)
		return
	}
	c.JSON(http.StatusOK, taskDTOToJSON(dto))
}

func (h *TaskHandler) Delete(c *gin.Context) {
	userID, ok := h.resolveUserID(c)
	if !ok {
		return
	}
	id, err := parseUintParam(c, "id")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid_id"})
		return
	}
	if err := h.tasks.Delete(c.Request.Context(), id, userID); err != nil {
		renderServiceError(c, err)
		return
	}
	c.Status(http.StatusNoContent)
}

// CONTRACT.md セクション5.1のJSON形状(スネークケース)に合わせる
// bffのtaskV1DTOはこのケースでアンマーシャルするため、camelCaseで返すと
// finished_on/created_at/updated_atがすべて空文字になる(統合時に発覚した不整合)
func taskDTOToJSON(dto service.TaskDTO) gin.H {
	labels := make([]gin.H, 0, len(dto.Labels))
	for _, l := range dto.Labels {
		labels = append(labels, gin.H{"id": l.ID, "name": l.Name})
	}
	return gin.H{
		"id":          dto.ID,
		"name":        dto.Name,
		"description": dto.Description,
		"status":      dto.Status,
		"finished_on": dto.FinishedOn.Format("2006-01-02"),
		"labels":      labels,
		"created_at":  dto.CreatedAt.Format(time.RFC3339),
		"updated_at":  dto.UpdatedAt.Format(time.RFC3339),
	}
}

func queryIntDefault(c *gin.Context, name string, def int) int {
	v, ok := parseUintQuery(c, name)
	if !ok {
		return def
	}
	return int(v)
}

File: backend-js-ts-express/src/rest/task.ts
// backend-js/src/rest/task.jsの型付き移植。ロジックは変更していない。

import type { Request, Response, NextFunction, RequestHandler } from 'express';
import * as db from '../db';
import * as errors from '../error';
import { authenticate } from './authenticate';
import { validateTaskInput, todayUtcIso, statusToString, statusFromString } from '../model';
import type { AppState, Task, TaskInput } from '../types';

function wrap(fn: (req: Request, res: Response) => Promise<void>): RequestHandler {
  return (req: Request, res: Response, next: NextFunction) => {
    fn(req, res).catch(next);
  };
}

function parseId(raw: string): number {
  if (!/^\d+$/.test(raw)) throw errors.invalidId();
  return Number(raw);
}

// backend(Go)のparseUintListQueryと同じく、パースに失敗した要素は無視して続行する
export function parseLabelIds(raw: string | undefined): number[] {
  if (!raw) return [];
  return raw
    .split(',')
    .map((p) => p.trim())
    .filter((p) => /^\d+$/.test(p))
    .map((p) => Number(p));
}

// CONTRACT.mdセクション5.1のJSON形状(スネークケース)。
// 【backend(Go)の実際の挙動に合わせた既知の差異】taskDTOToJSON(backend/internal/handler/v1/task.go)は
// user_idをレスポンスに含めていない。
export function taskToJson(task: Task): Record<string, unknown> {
  return {
    id: task.id,
    name: task.name,
    description: task.description,
    status: statusToString(task.status),
    finished_on: task.finishedOn,
    labels: task.labels.map((l) => ({ id: l.id, name: l.name })),
    created_at: `${task.createdAt.replace(' ', 'T')}+00:00`,
    updated_at: `${task.updatedAt.replace(' ', 'T')}+00:00`,
  };
}

export function list(state: AppState) {
  return wrap(async (req, res) => {
    const userId = await authenticate(state, req);
    const q = req.query as Record<string, string | undefined>;

    let status: number | null = null;
    if (q.status) {
      status = statusFromString(q.status);
      if (status === null) throw errors.invalidStatus();
    }
    const labelIds = parseLabelIds(q.label_ids);
    const limit = q.limit !== undefined ? parseInt(q.limit, 10) : 20;
    const offset = q.offset !== undefined ? parseInt(q.offset, 10) : 0;

    const { tasks, total } = await db.listTasksOffset(
      state.pool,
      userId,
      q.name || '',
      status,
      labelIds,
      q.sort || '',
      limit,
      offset,
    );
    res.json({ tasks: tasks.map(taskToJson), total, limit, offset });
  });
}

export function get(state: AppState) {
  return wrap(async (req, res) => {
    const userId = await authenticate(state, req);
    const id = parseId(req.params.id!);
    const task = await db.getTask(state.pool, id, userId);
    if (!task) throw errors.notFound();
    res.json(taskToJson(task));
  });
}

// backend(Go)のtaskRequestBody(binding:"required")と同じ「空文字も必須違反」の扱い
export function toTaskInput(body: Record<string, unknown>): TaskInput {
  const name = (body.name as string) ?? '';
  const status = (body.status as string) ?? '';
  const finishedOn = (body.finished_on as string) ?? '';
  if (!name || !status || !finishedOn) throw errors.invalidRequest();
  if (!/^\d{4}-\d{2}-\d{2}$/.test(finishedOn)) throw errors.invalidFinishedOn();

  // カレンダー上有効な日付かも確認する(例: 2026-02-30を弾く)
  const [y, m, d] = finishedOn.split('-').map(Number) as [number, number, number];
  const check = new Date(Date.UTC(y, m - 1, d));
  if (check.getUTCFullYear() !== y || check.getUTCMonth() !== m - 1 || check.getUTCDate() !== d) {
    throw errors.invalidFinishedOn();
  }

  const labelIds = Array.isArray(body.label_ids) ? (body.label_ids as unknown[]).map(Number) : [];
  return { name, description: (body.description as string) ?? null, status, finishedOn, labelIds };
}

export function create(state: AppState) {
  return wrap(async (req, res) => {
    const userId = await authenticate(state, req);
    const input = toTaskInput(req.body || {});

    const result = validateTaskInput(input, todayUtcIso());
    if (result.error !== undefined) throw errors.validationError(result.error);

    const id = await db.createTask(state.pool, userId, input, result.status);
    const task = await db.getTask(state.pool, id, userId);
    res.status(201).json(taskToJson(task!));
  });
}

export function update(state: AppState) {
  return wrap(async (req, res) => {
    const userId = await authenticate(state, req);
    const id = parseId(req.params.id!);
    const input = toTaskInput(req.body || {});

    const result = validateTaskInput(input, todayUtcIso());
    if (result.error !== undefined) throw errors.validationError(result.error);

    const updated = await db.updateTask(state.pool, id, userId, input, result.status);
    if (!updated) throw errors.notFound();
    const task = await db.getTask(state.pool, id, userId);
    res.json(taskToJson(task!));
  });
}

export function remove(state: AppState) {
  return wrap(async (req, res) => {
    const userId = await authenticate(state, req);
    const id = parseId(req.params.id!);
    const deleted = await db.deleteTask(state.pool, id, userId);
    if (!deleted) throw errors.notFound();
    res.status(204).send();
  });
}
